### PR TITLE
CA-396188 [Terraform Provider] Unable to read VM platform cores-per-socket

### PR DESCRIPTION
**Pull request by acefei**
```
source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkDataSource
--- PASS: TestAccNetworkDataSource (0.78s)
=== RUN   TestAccVlanResource
--- PASS: TestAccVlanResource (3.60s)
=== RUN   TestAccNICDataSource
--- PASS: TestAccNICDataSource (1.19s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (0.90s)
=== RUN   TestAccSnapshotResource
--- PASS: TestAccSnapshotResource (6.30s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (0.75s)
=== RUN   TestAccNFSResource
--- PASS: TestAccNFSResource (9.21s)
=== RUN   TestAccSRResourceLocal
--- PASS: TestAccSRResourceLocal (4.80s)
=== RUN   TestAccSRResourceShared
--- PASS: TestAccSRResourceShared (11.46s)
=== RUN   TestAccVDIResource
--- PASS: TestAccVDIResource (16.09s)
=== RUN   TestAccVMDataSource
--- PASS: TestAccVMDataSource (2.37s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (4.30s)
=== RUN   TestAccLinuxVMResource
--- PASS: TestAccLinuxVMResource (2.34s)
PASS
ok      terraform-provider-xenserver/xenserver  64.116s
```

### CODE REVIEW

File: xenserver/vm_utils.go [hunk 1]
```diff
@@ -472,10 +469,13 @@ func setOtherConfigFromPlan(ctx context.Context, session *xenapi.Session, vmRef
 		tflog.Debug(ctx, "-----> setOtherConfig key: "+key+" value: "+value)
 	}
 
-	originalTFOtherConfigKeys := vmOtherConfig["tf_other_config_keys"]
+	originTFOtherConfigKeys, ok := vmOtherConfig["tf_other_config_keys"]
+	if !ok {
```
Comment (xueqingz) [inline xenserver/vm_utils.go:473]: originTFOtherConfigKeys = "" if !ok, don't need this.

File: xenserver/vm_utils.go [hunk 2]
```diff
@@ -735,7 +725,17 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
 		}
 	}
 
-	if !data.CorePerSocket.IsUnknown() {
+	platform := vmRecord.Platform
+	if data.CorePerSocket.IsUnknown() {
+		// if user doesn't set cores-per-socket and it is not found in template, set it to VCPUs num as the default value
+		if _, ok := platform["cores-per-socket"]; !ok {
+			platform["cores-per-socket"] = data.VCPUs.String()
```
Comment (xueqingz) [inline xenserver/vm_utils.go:732]: why use "data.VCPUs.String()" ?

File: xenserver/vm_utils.go [hunk 3]
```diff
@@ -745,107 +745,72 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
 
-func updateBootOrder(session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
+func updateBootOrder(session *xenapi.Session, vmRef xenapi.VMRef, vmRecord xenapi.VMRecord, plan vmResourceModel) error {
 	var err error
-	bootOrder := plan.BootOrder.ValueString()
-	if bootOrder == "" {
+
+	// don't set boot order if it is unknown, using the default value from the template
+	if plan.BootOrder.IsUnknown() {
 		return err
```
Comment (xueqingz) [inline xenserver/vm_utils.go:758]: easy miss-understanding here, just return nil.

File: xenserver/vm_utils.go [hunk 4]
```diff
@@ -745,107 +745,72 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
-			return fmt.Errorf("%d cores could not fit to %d cores-per-socket topology", vcpus, socketInt)
-		}
 	}
 
 	return nil
 }
 
-func updateBootOrder(session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
+func updateBootOrder(session *xenapi.Session, vmRef xenapi.VMRef, vmRecord xenapi.VMRecord, plan vmResourceModel) error {
 	var err error
```
Comment (xueqingz) [inline xenserver/vm_utils.go:754]: remove "var err error" and use 
err := xenapi.VM.SetHVMBootParams(session, vmRef, hvmBootParams)
in follow may be better?

File: xenserver/vm_utils.go (hunk 3 above)
Comment (acefei) [reply xenserver/vm_utils.go:758]: I don't think so, there is a comment to elabarate the reason.

File: xenserver/vm_utils.go (hunk 1 above)
Comment (acefei) [reply xenserver/vm_utils.go:473]: I noticed you tested it , so agree,  I will remove this logic.

File: xenserver/vm_utils.go (hunk 2 above)
Comment (acefei) [reply xenserver/vm_utils.go:732]: It follows the logic of XenCenter, when changing vcpu num, the cores_per_socket would be changed as the same as vCPUs.

File: xenserver/vm_utils.go (hunk 4 above)
Comment (acefei) [reply xenserver/vm_utils.go:765]: `var err error` is for the guard clauses `return err`

**Review by xihuan-citrix — APPROVED** [review]

File: xenserver/vm_utils.go [hunk 5]
```diff
@@ -754,97 +756,55 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
 	}
 
-	return err
-}
-
-func setPlatform(session *xenapi.Session, vmRef xenapi.VMRef, params map[string]string) error {
-	originalValue, err := xenapi.VM.GetPlatform(session, vmRef)
-	if err != nil {
-		return errors.New(err.Error())
+	secureBoot, ok := vmRecord.Platform["secureboot"]
```
Comment (xueqingz) [inline xenserver/vm_utils.go:788]: Does the template always have Platform["secureboot"] ?

**Review by xueqingz — APPROVED** [review]

File: xenserver/vm_utils.go (hunk 5 above)
Comment (acefei) [reply xenserver/vm_utils.go:788]: yes, Linux and Windows template both include secureboot.
https://github.com/xenserver/guest-templates-json/blob/master/json/base-hvmlinux.json
https://github.com/xenserver/guest-templates-json/blob/master/json/base-windows.json

**Pull request merged**